### PR TITLE
[NativeAnimated][iOS] Value offsets

### PR DESCRIPTION
**Pull request by ryangomba**
This diff adds support for value offsets on iOS. It separates out code originally submitted in #9048.

Test plan (required)

Set up an animation with an offset, and `useNativeModule: true`. Compare results with `useNativeModule: false`.

### CODE REVIEW

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 1]
```diff
@@ -725,16 +725,24 @@ class AnimatedValue extends AnimatedWithChildren {
    * things like the start of a pan gesture.
    */
   setOffset(offset: number): void {
-    this._offset = offset;
+    if (this.__isNative) {
+      NativeAnimatedAPI.setAnimatedNodeOffset(this.__getNativeTag(), offset);
+    } else {
```
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: I think we should still set `this._offset` even if it is native.

File: Libraries/Animated/src/AnimatedImplementation.js [hunk 2]
```diff
@@ -725,16 +725,24 @@ class AnimatedValue extends AnimatedWithChildren {
...
   /**
    * Merges the offset value into the base value and resets the offset to zero.
    * The final output of the value is unchanged.
    */
   flattenOffset(): void {
-    this._value += this._offset;
-    this._offset = 0;
+    if (this.__isNative) {
+      NativeAnimatedAPI.flattenAnimatedNodeOffset(this.__getNativeTag());
+    } else {
```
Comment (janicduplessis) [inline Libraries/Animated/src/AnimatedImplementation.js]: Same here, we should try to keep the JS values in sync even if it is native driven.

**Comment by javache** [comment]
cc @ritzau

**Comment by mkonicek** [comment]
@foghina Are you or @ritzau the best reviewer for this one?

**Comment by mkonicek** [comment]
Ah actually this is in 'needs-revision' state thanks to Janic's review :)

**Comment by ryangomba** [comment]
Sorry for the delay, folks. @janicduplessis - I've addressed your comments.

File: Libraries/NativeAnimation/Nodes/RCTValueAnimatedNode.m [hunk 3]
```diff
@@ -9,8 +9,35 @@
...
+                     config:(NSDictionary<NSString *, id> *)config
+{
+  if (self = [super initWithTag:tag config:config]) {
+    _offset = [self.config[@"offset"] floatValue];
+    _value = [self.config[@"value"] floatValue];
+  }
+  return self;
+}
+
+- (void)flattenOffset {
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/Nodes/RCTValueAnimatedNode.m]: Nit: new line

File: Libraries/NativeAnimation/Nodes/RCTValueAnimatedNode.m [hunk 4]
```diff
@@ -9,8 +9,35 @@
...
+  _value += _offset;
+  _offset = 0;
+}
+
+- (CGFloat)value {
```
Comment (janicduplessis) [inline Libraries/NativeAnimation/Nodes/RCTValueAnimatedNode.m]: Same here

**Comment by facebook-github-bot-5** [comment]
@ryangomba updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9627/files/6baaca8aaa49180675a17ce14afcdb65748d0a1d..741fd0be73d3a37b4076e8bbc064092a0e97eb26)

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-3** [comment]
Thanks for importing. If you are a Facebook employee, you can view this diff [on Phabricator](https://phabricator.intern.facebook.com/D3924410).

**Pull request closed without merging**